build/common.sh

```diff
@@ -566,14 +533,38 @@ function kube::build::ensure_data_container() {
     # We want this to run as root to be able to chown, so non-root users can
     # later use the result as a data container.  This run both creates the data
     # container and chowns the GOPATH.
+    #
+    # The data container creates volumes for all of the directories that store
+    # intermediates for the Go build. This enables incremental builds across
+    # Docker sessions. The *_cgo paths are re-compiled versions of the go std
+    # libraries for true static building.
     local -ra docker_cmd=(
       "${DOCKER[@]}" run
+      --volume "${REMOTE_ROOT}"   # white-out the whole output dir
+      --volume /usr/local/go/pkg/linux_386_cgo
+      --volume /usr/local/go/pkg/linux_amd64_cgo
+      --volume /usr/local/go/pkg/linux_arm_cgo
+      --volume /usr/local/go/pkg/linux_arm64_cgo
+      --volume /usr/local/go/pkg/linux_ppc64le_cgo
+      --volume /usr/local/go/pkg/darwin_amd64_cgo
+      --volume /usr/local/go/pkg/darwin_386_cgo
+      --volume /usr/local/go/pkg/windows_amd64_cgo
+      --volume /usr/local/go/pkg/windows_386_cgo
       --name "${KUBE_BUILD_DATA_CONTAINER_NAME}"
       --hostname "${HOSTNAME}"
-      --volume "${REMOTE_OUTPUT_ROOT}"   # white-out the whole output dir
       --volume "${REMOTE_OUTPUT_GOPATH}" # make a non-root owned mountpoint
       "${KUBE_BUILD_IMAGE}"
-      chown -R $(id -u).$(id -g) "${REMOTE_OUTPUT_ROOT}"
+      chown -R $(id -u).$(id -g)
+        "${REMOTE_ROOT}"
+        /usr/local/go/pkg/linux_386_cgo
```

Comment:
I meant just to chmod/chown the whole thing, not to volume mount the whole
thing.

On Mon, Aug 22, 2016 at 2:24 PM, Joe Beda <notifications@github.com> wrote:

> In build/common.sh
> <https://github.com/kubernetes/kubernetes/pull/30787#discussion_r75761703>
> :
>
> >        --volume "${REMOTE_OUTPUT_GOPATH}" # make a non-root owned mountpoint
> >        "${KUBE_BUILD_IMAGE}"
> > -      chown -R $(id -u).$(id -g) "${REMOTE_OUTPUT_ROOT}"
> > +      chown -R $(id -u).$(id -g)
> > +        "${REMOTE_ROOT}"
> > +        /usr/local/go/pkg/linux_386_cgo
>
> The compiler itself is under /usr/local/go/pkg. If we make that a volume
> it gets whited out. Not sure what else to do there.
>
> We could do something where we move it aside when building the image and
> then copy it back as part of the chown step, but that seems complicated too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30787/files/c0ba418d1f5f07f74ae8e9aa392e7ddbcfa0a144#r75761703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVEMQP4zlhDmbqAsqoZD3z7Rl7Y3Yks5qihN5gaJpZM4Jmt7e>
> .
>